marlowe-runtime/sync/Language/Marlowe/Runtime/Sync/Database/PostgreSQL/GetIntersectionForContract.hs

```diff
@@ -0,0 +1,77 @@
+{-# LANGUAGE QuasiQuotes #-}
+
+module Language.Marlowe.Runtime.Sync.Database.PostgreSQL.GetIntersectionForContract
+  where
+
+import Data.ByteString (ByteString)
+import Data.Int (Int64)
+import Data.Maybe (listToMaybe)
+import qualified Data.Vector as V
+import Hasql.TH (vectorStatement)
+import qualified Hasql.Transaction as T
+import Language.Marlowe.Runtime.ChainSync.Api (BlockHeader(..), BlockHeaderHash(..), TxId(..), TxOutRef(..))
+import Language.Marlowe.Runtime.Core.Api (ContractId(ContractId), MarloweVersion(..), SomeMarloweVersion(..))
+
+getIntersectionForContract :: ContractId -> [BlockHeader] -> T.Transaction (Maybe (BlockHeader, SomeMarloweVersion))
+getIntersectionForContract _ [] = pure Nothing
+getIntersectionForContract (ContractId TxOutRef{..}) (b : bs) = do
+  serverBlocks <- fmap decodeBlock . V.toList <$> T.statement params
+    [vectorStatement|
+      WITH contractId (txId, txIx) AS
+        ( SELECT $1 :: bytea, $2 :: smallint
+        )
+      SELECT
+        block.slotNo :: bigint,
+        block.id :: bytea,
+        block.blockNo :: bigint
+      FROM marlowe.createTxOut
+      JOIN marlowe.block
+        ON block.id = createTxOut.blockId
+      JOIN contractId USING (txId, txIx)
+      WHERE block.rollbackToSlot IS NULL
+      UNION
+      SELECT DISTINCT
```

Comment:
Actually, isn't `DISTINCT` redundant because `UNION` eliminates duplicates?

Reply:
Also, it might be worthwhile adding a comment that the purpose of the unions is to find creation, applications, and withdrawals involving the contract.

Reply:
I can add that comment

Reply:
Does `UNION` eliminate duplicates within the sub queries, or only between the two queries?